File: Queue.cpp
#include <iostream>
#include <vector>

using namespace std;

struct Node{
	int value;
	Node* next;
	Node(int val):value(val), next(NULL) {}
};

class Queue{
private:
	Node* top;
	Node* buttom;
	int size;

public:
	Queue(){
		this->top = NULL;
		this->buttom = NULL;
		this->size = 0;
	}

	void peek(){
		if(this->size == 0) {cout <<"NULL\n"; return;}
		cout <<buttom->value;
	}

	void enqueue(int value){
		Node* newNode = new Node(value);
		if(this->size == 0) {
			this->buttom = newNode;
			this->top = this->buttom;
		}else{
			this->top->next = newNode;
			this->top = newNode;
		}
		this->size++;
	}
	void dequeue(){
		if(this->size == 0) return;
		else{ //Deallocate memeory here.
			this->buttom = this->buttom->next;
		}
		this->size--;
	}

	int Size(){
		return this->size;
	}

	void PrintQueue(){
		Node* temp = this->buttom;
		while(temp != NULL){
			cout <<temp->value <<" ";
			temp = temp->next;
		}
		cout <<"\n";
	}
};

int main(){

	Queue q;
	q.enqueue(1);
	q.enqueue(2);
	q.enqueue(3);
	q.enqueue(4);
	q.dequeue();

	cout << q.Size()<<endl;

	q.PrintQueue();


	return 0;
}

File: Queue.js
class Node{
  constructor(value){
    this.value = value;
    this.next = null;
  }
}

class Queue{
	constructor(){
		this.first = null;
		this.last = null;
		this.length = 0;
	}
	peek(){
		return this.last;
	}
	enqueue(value){
		const newNode = new Node(value);
		if(this.length == 0){
			this.last = newNode;
			this.first = this.last;
		}else{
			this.first.next = newNode;
			this.first = newNode;
		}
		this.length++;
	}
	dequeue(){
		if(this.length == 0) return null;
		this.last = this.last.next;
		this.length--;
	}
	size(){
		return this.length;
	}
	printQueue(){
		let temp = this.last;
		const arri = [];
		while(temp != null){
			arri.push(temp.value);
			temp = temp.next;
		}
		return arri;
	}
}
